README.txt
= ShippingCalc

* http://github.com/davertron/shipping_calc/
* mailto:davertron@gmail.com

== DESCRIPTION:

Shipping Calculator written in Ruby to get quick quotes from DHL, UPS and Freight Carriers.
We hope to support FedEx in the near future.	

== FEATURES/PROBLEMS:

- Current version does not support FedEx

== SYNOPSIS:

You can find an example of each carrier's API under the /examples directory.

A simple DHL example:

	require 'rubygems'
	require 'shipping_calc'

	include ShippingCalc

	opts = {
	  :api_user => "your_user", 
	  :api_password => "your_pwd",
	  :shipping_key => "your_key",
	  :account_num => "your_accnt",
	  :date => Time.now, 
	  :service_code => "E", # check the DHL docs to find out what this means
	  :shipment_code => "P", # check the DHL docs to find out what this means
	  :weight => 34, # weight in lbs
	  :to_zip => 10001,
	  :to_state => "NY"
	}

	d = DHL.new
	q = d.quote(opts)
	puts q

== REQUIREMENTS:

* You must obtain all the DHL ShipIt data (user, password, key and account) from http://www.dhl-usa.com/TechTools/detail/TTDetail.asp?nav=TechnologyTools/Shipping/OwnSoln

== INSTALL:

* sudo gem install shipping-calc

== TEST:

To run the DHL tests you'll need to have a .dhl_info.yml file in your home directory with your auth info like this: 
    ~/.dhl_info.yml

    api_user: your_user
    api_password: your_password
    shipping_key: your_key
    account_num: your_accnt_num

To run the UPS tests, you will need a similar setup, with a .ups_info.yml in
your home directory with info like this:
    ~/.ups_info.yml
    
    api_user: your_user
    api_password: your_password
    shipping_key: your_key

This is necessary only for the tests. When using the library, you can pass this
information in to the UPS/DHL/Freight classes in any fashion you like.

== LICENSE:

Copyright (c) 2009 David Davis

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 0.2.0 / 2009-11-04

Added support for UPS shipping quotes.

=== 0.1.7 / 2008-06-24

Improvements for DHL:
  * Calculations now handle and print descriptive errors when a quote fails.
  * Fixed a bug with the default ship date not being set
  * Tests should now un in Mac OS X.

=== 0.1.6 / 2008-05-29

* Fixed a bug in FreightQuote where the origin and destination zip code would
  some times be the same.

=== 0.1.5 / 2008-04-20

* DHL quotes now throw an exception if the DHL's sytem is online (like it is
  right now!).

=== 0.1.4 / 2008-04-12

* Fixed some tests and documentation

=== 0.1.3 / 2008-03-17

* FreightQuote support improved

Now you can query FQ's API with information for:

  * Liftgates
  * Loading docks
  * Different types of shipping and receiving locations
  * Inside deliveries 

* Minor bugfixes

=== 0.1.2

* Freight quote now supports FQ's classes. With this the shipment's
dimensions are optional

=== 0.1.1
* Usage examples for DHL and FreightQuote in the /examples directory.

=== 0.1.0
* Proper support for FreightQuote quotes.
FreightQuote uses an API that calculates the cost for a freight shipping
accross several carriers and returns a list of their rates. ShippingCalc will
now be able to query them and return a hash like {carrier_name => rate} so
you can pick whatever's more useful to you.

=== 0.0.2 / 2008-03-13

* Adds basic support for FreightQuote.

=== 0.0.1 / 2008-03-09

* Initial buggy release

  * Don't try this at home!



lib/shipping_calc.rb
# Copyright (c) 2008 Federico Builes

# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# 'Software'), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:

# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

$:.unshift(File.dirname(__FILE__))

require 'rubygems'
require File.dirname(__FILE__) + "/shipping_calc/dhl"
require File.dirname(__FILE__) + "/shipping_calc/ups"
require File.dirname(__FILE__) + "/shipping_calc/freight_quote"
module ShippingCalc

  VERSION = "0.2.0" 

  US_STATES =  ['AK', 'AL', 'AR', 'AZ', 'CA', 'CO', 'CT', 'DC',
               'DE', 'FL', 'GA', 'HI', 'IA', 'ID', 'IL', 'IN',
               'KS', 'KY', 'LA', 'MA', 'MD', 'ME', 'MI', 'MN',
               'MO', 'MS', 'MT', 'NC', 'ND', 'NE', 'NH', 'NJ',
               'NM', 'NV', 'NY', 'OH', 'OK', 'OR', 'PA', 'RI',
               'SC', 'SD', 'TN', 'TX', 'UT', 'VA', 'VT', 'WA',
               'WI', 'WV'] 


  class ShippingCalcError < StandardError
  end
end


examples/ups_example.rb
require 'rubygems'
require File.dirname(__FILE__) + '/../lib/shipping_calc'
require 'yaml'
require 'rbconfig'
include ShippingCalc

# This example requires you to have a .ups_info.yml file in your home dir to
# gather the login data. Check the Test secion in README.txt for more information.
begin
  os = Config::CONFIG["host_os"]
  if os =~ /darwin/
    auth_info = YAML.load_file("/Users/#{ENV["USER"]}/.ups_info.yml")
    else      
    auth_info = YAML.load_file("/home/#{ENV["USER"]}/.ups_info.yml")
  end
rescue Exception
  print "You don't have a .ups_info.yml file in your home directory. Please
read the \"Test\" section in README.txt.\n"
  exit
end

api_user = auth_info["api_user"]
api_pwd = auth_info["api_password"]
api_key = auth_info["shipping_key"]

opts = { 
  :api_user => api_user,
  :api_password => api_pwd,
  :shipping_key => api_key,
  :date => Time.now,
  :from_city => 'South Burlington',
  :from_state => 'VT',
  :from_country => 'US',
  :from_zip => '05403',
  :to_city => 'Shelburne',
  :to_zip => '05482',
  :to_state => 'VT',
  :to_country => 'US',
  :pickup_code => '01', # Daily Pickup, see lib/shipping_calc/ups.rb codes
  :package_code => '02', # UPS Letter, see lib/shipping_calc for details
  :weight_uom => 'LBS', # Unit of measure for package weight
  :package_weights => ['10']
}

u = UPS.new
q = u.quote(opts)
puts "Quote:"
q.each do |quote|
    puts "#{quote[0]} : #{quote[1]}"
end


lib/shipping_calc/dhl.rb
# Copyright (c) 2008 Federico Builes

# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# 'Software'), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:

# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

require 'rexml/document'
require 'net/http'
require 'net/https'
require 'uri'
include REXML

module ShippingCalc
  # Current version on their website is 1.0 and can be found in:
  # https://eCommerce.airborne.com/ApiLandingTest.asp . To get full access to
  # all their stuff you have to make sure they certify your application
  # against their live platform tests. The test bed should be enough to get
  # simple calculations.
  # Currently, only shipments made inside the US are available.

  class DHL

    # Obtains an estimate quote from the DHL site. 
    # <tt>params</tt> is a hash with all the settings for the shipment. They are:
    #
    # :*api_user*:: API access username, provided by DHL.
    # :*api_password*:: API access password, provided by DHL.
    # :*shipping_key*:: API shipping key, provided by DHL.
    # :*international_shipping_key*:: API shipping key, provided by DHL, used
      # for international quotes.
    # :*account_num*:: Account number, provided by DHL.
    # :*date*:: Date for the shipping in format YYYY-MM-DD (defaults to Time.now).
    # :*service_code*:: Service code defined in Rate Estimate Specification(E, N, S, G). 1030 and SAT are not supported yet. Defaults to G (ground service).
    # :*shipment_code*:: ShipmentType code defined in the Rate Estimate Specification. "P" for Package or "L" for Letter. Defaults to "P".
    # :*weight*:: Order's weight. If the shipment code is a "L" (letter) then the weight will be 0.
    # :*to_street*:: Recipient's street
    # :*to_zip*:: Recipient's zip code.
    # :*to_country*:: Recipient's country.
    # :*to_state*:: Recipient's state.
    # :*international*:: Whether or not the quote is international or not

    def quote(params)
      @international = params[:international]
      @xml = Document.new
      @xml << XMLDecl.new("1.0' encoding='UTF-8")
      raise ShippingCalcError.new("Invalid parameters") if params.nil?
      raise ShippingCalcError.new("Missing shipping parameters") unless params.keys.length >= 10
      auth(params[:api_user], params[:api_password])
      rate_estimate(params)
      request
    end

    private 

    # DHL gets the quote in 2 steps, the first one is authentication. This
    # generates the XML for that.
    def auth(api_user, api_password)
      # <eCommerce action="Request" version="1.1">
      #   <Requestor>
      #     <ID>username</ID>
      #     <Password>password</Password>
      #   </Requestor>
      # </eCommerce>
      ecommerce = Element.new "eCommerce"
      ecommerce.attributes["action"] = "Request"
      ecommerce.attributes["version"] = "1.1"

      user = Element.new "Requestor"
      user_id = Element.new "ID"
      user_id.text = api_user 

      user_pwd = Element.new "Password"
      user_pwd.text = api_password 

      user << user_id
      user << user_pwd

      ecommerce << user
      @xml << ecommerce
    end

    # After having the auth message ready, we create the RateEstimate request.
    #     shipping_key: API shipping key, provided by DHL.
    #     account_num: Account number, provided by DHL.
    #     date: Date for the shipping. Must be a Ruby "Time" object.
    #     service_code: Service code defined in Rate Estimate Specification
    #     (E, N, S, G). 1030 and SAT are not supported yet. Defaults to G
    #     (ground service).
    #     shipment_code: ShipmentType code defined in the Rate Estimate
    #     Specification. "P" for Package or "L" for Letter. Defaults to "P".
    #     weight: Order's weight. If the shipment code is a "L" (letter) then
    #     the weight will be 0.
    #     to_street: Recipient's street address.
    #     to_zip: Recipient's zip code.
    #     to_country: Recipient's country. Not used, currently DHL only supports US.
    #     to_state: Recipient's state.
    #     international: Whether you're getting an international quote or not
    def rate_estimate(params)
      if @international
        shipment = Element.new 'IntlShipment'
        shipment.attributes['action'] = 'RateEstimate'
        shipment.attributes['version'] = '1.0'
        
        credentials = Element.new 'ShippingCredentials'
        key = Element.new 'ShippingKey'
        key.text = params[:international_shipping_key]
        account = Element.new 'AccountNbr'
        account.text = params[:account_num]

        credentials << key
        credentials << account
        shipment << credentials

        shipment_details = Element.new 'ShipmentDetail'
        ship_date = Element.new 'ShipDate'
        ship_date.text = date(params[:date])
        shipment_details << ship_date

        service = Element.new 'Service'
        service_code = Element.new 'Code'
        service_code.text = service_code(params[:service_code])
        service << service_code
        shipment_details << service

        shipment_type = Element.new 'ShipmentType'
        type_code = Element.new 'Code'
        type_code.text = shipment_code(params[:shipment_code])
        shipment_type << type_code
        shipment_details << shipment_type

        weight = Element.new 'Weight'
        weight.text = weight(params[:weight], 'P')
        shipment_details << weight

        content_description = Element.new 'ContentDesc'
        content_description.text = 'Computer Parts'
        shipment_details << content_description

        shipment << shipment_details

        dutiable = Element.new 'Dutiable'
        dutiable_flag = Element.new 'DutiableFlag'
        dutiable_flag.text = 'Y'
        dutiable << dutiable_flag

        customs_value = Element.new 'CustomsValue'
        customs_value.text = '100'
        dutiable << customs_value
        shipment << dutiable

        billing = Element.new 'Billing'
        party = Element.new 'Party'
        party_code = Element.new 'Code'
        party_code.text = 'S'
        party << party_code
        billing << party
        duty_payment_type = Element.new 'DutyPaymentType'
        duty_payment_type.text = 'S'
        billing << duty_payment_type
        shipment << billing

        receiver = Element.new 'Receiver'
        address = Element.new 'Address'
        street = Element.new 'Street'
        street.text = params[:to_street]
        address << street
        city = Element.new 'City'
        city.text = params[:to_city]
        address << city
        state = Element.new 'State'
        state.text = params[:to_state]
        address << state
        country = Element.new 'Country'
        country.text = params[:to_country]
        address << country
        postal_code = Element.new 'PostalCode'
        postal_code.text = params[:to_zip]
        address << postal_code
        receiver << address

        shipment << receiver
      else
        # <Shipment action="RateEstimate" version="1.0">
        shipment = Element.new "Shipment"
        shipment.attributes["action"] = "RateEstimate"
        shipment.attributes["version"] = "1.0"

        # <ShippingCredentials>
        #   <ShippingKey>key</ShippingKey> 
        #   <AccountNbr>number</AccountNbr>
        # </ShippingCredentials>
        credentials = Element.new "ShippingCredentials"

        key = Element.new "ShippingKey"
        key.text = params[:shipping_key] 
        account = Element.new "AccountNbr"
        account.text = params[:account_num] 

        credentials << key
        credentials << account
        shipment << credentials

        # <ShipmentDetail>
        #   <ShiptDate>date</ShipDate>
        shipment_detail = Element.new "ShipmentDetail"
        ship_date = Element.new "ShipDate"   
        ship_date.text = date(params[:date])
        shipment_detail << ship_date

        # TODO: Implement SAT and 1030 services

        #   <Service>
        #     <Code>code</Code>
        #   </Service>
        service = Element.new "Service"
        service_code = Element.new "Code"
        service_code.text = service_code(params[:service_code])
        service << service_code

        shipment_detail << service

        #   <ShipmentType>
        #     <Code>code</Code>
        #   </ShipmentType>
        shipment_type = Element.new "ShipmentType"
        shipment_type_code = Element.new "Code"
        shipment_type_code.text = shipment_code(params[:shipment_code])
        shipment_type << shipment_type_code
        shipment_detail << shipment_type

        #   <Weight>weight</Weight>
        weight = Element.new "Weight"
        weight.text = weight(params[:weight], params[:shipment_code])

        # </ShipmentDetail>      
        shipment_detail << weight

        shipment << shipment_detail

        # <Billing>
        #   <Party>
        #     <Code>S</Code>
        #   </Party>
        # </Billing>
        billing = Element.new "Billing"
        billing_party = Element.new "Party"
        billing_party_code = Element.new "Code"

        # Since we're just doing some quick calulations we don't want to be
        # worrying about who's gonna send the package. Just make the calulations 
        # assuming the sender pays for the shipping.
        billing_party_code.text = "S"

        billing << billing_party << billing_party_code

        shipment << billing

        # <Receiver>
        #   <Address>
        #     <State>state</State>
        #     <Country>country</Country>
        #     <PostalCode>code</PostalCode>
        #   </Address>
        # </Receiver>
        receiver = Element.new "Receiver"
        receiver_addr = Element.new "Address"
        receiver_state = Element.new "State"
        receiver_country = Element.new "Country"
        receiver_zipcode = Element.new "PostalCode"

        receiver_state.text = state(params[:to_state])
        receiver_country.text = "US"
        receiver_zipcode.text = zip_code(params[:to_zip])

        receiver_addr << receiver_state
        receiver_addr << receiver_country
        receiver_addr << receiver_zipcode
        receiver << receiver_addr

        shipment << receiver
      end

      root = @xml.elements["eCommerce"]
      root.add shipment
    end

    # Sends the request to the web server and returns the response.
    def request
      #server = Net::HTTP.new("eCommerce.airborne.com", 443)
      server = Net::HTTP.new("eCommerce.airborne.com", 80)
      #path = "/ApiLandingTest.asp"
      path = "/ApiLanding.asp"
      data = @xml.to_s
      File.open('/home/david/Desktop/dhl_intl.xml', 'w') do |f|
        f.puts data
      end
      headers = { "Content-Type" => "text/xml"}
      #server.use_ssl = true
      resp = server.post(path, data, headers)
      price = parse_response(resp.body)
    end

    # Parses the server's response. Currently, it only returns the estimate
    # value of the shipping.
    def parse_response(resp)
      doc = Document.new(resp)

      find_error_and_raise(doc) if errors_exist?(doc)

      if @international
        result =  doc.elements["//IntlShipment/Result/Desc"].text 
      else
        result =  doc.elements["//Shipment/Result/Desc"].text 
      end
      
      if result == "Shipment estimate successful."
        if @international
          doc.elements["//IntlShipment/EstimateDetail/RateEstimate/TotalChargeEstimate"].text.to_f
        else
          doc.elements["//Shipment/EstimateDetail/RateEstimate/TotalChargeEstimate"].text.to_f
        end
      else
        raise ShippingCalcError.new(doc.to_s)
      end
    end

    def errors_exist?(response)
      not response.elements["//Faults"].nil?
    end

    def find_error_and_raise(response)
    error_code = response.elements["//Code"]

        # Special Services and Additional Protection are not supported        
      case error_code.text.to_i
          when 1000..1009 then msg = "Shipment Headers"
          when 4000..4004 then msg = "ShippingKey"
          when 4007       then msg = "Account Number"
          when 4195..4198 then msg = "Account Number"
          when 4100..4106 then msg = "Shipment Date"
          when 4108..4117 then msg = "Service Type (Code)"
          when 4118..4122 then msg = "Shipment Type Code"
          when 4123..4124 then msg = "Weight"
          when 4128..4131 then msg = "Dimensions"
          when 4116       then msg = "Billing Party (Code)"
          when 4147       then msg = "Billing Party (Code)"
          when 4149..4152 then msg = "Billing Account Number"
          when 4164..4166 then msg = "Receiver City"
          when 4167       then msg = "Receiver State"
          when 4164..4166 then msg = "Receiver City"
          when 4169       then msg = "Receiver Country"
          when 4170..4176 then msg = "Receiver Postal Code"
          else                 msg = "API Request"
      end

      raise ShippingCalcError.new("DHL Error #{error_code.text}: Invalid #{msg}")
    end

    def date(date)
      date ||= Time.now
      if date.kind_of?(String) && date =~ /\d{4}-\d{2}-\d{2}/ # Suppose it's valid
        return date
      end

      if date.strftime("%A") == "Sunday"    
        (date + 86400).strftime("%Y-%m-%d") # DHL doesn't ship on Sundays, add 1 day.
      else
        date.strftime("%Y-%m-%d")
      end
    end

    def shipment_code(code)
      ['P', 'L', 'O'].include?(code) ? code : 'P'
    end

    def service_code(code)
      ['E', 'N', 'S', 'G', 'IE'].include?(code) ? code : 'G'
    end

    def weight(w, type)
      if type == 'L'
        '0'
      else
        (w > 0 && w <= 150) ? w.to_s : (raise ShippingCalcError.new('Invalid weight - Must be between 1 and 150 lbs.'))
      end
    end

    def state(s)
      valid_state?(s) ? s : (raise ShippingCalcError.new('Invalid state for recipient'))
    end

    def valid_state?(s)
      ShippingCalc::US_STATES.include?(s)
    end

    def zip_code(code)
      if code.class != Fixnum
        raise ShippingCalcError.new('Zip Code must be a number. Perhaps you are using a string?')
      end
      code.to_s =~ /\d{5}/ ? code.to_s : (raise ShippingCalcError.new('Invalid zip code for recipient'))
    end

  end
end


examples/dhl_intl_example.rb
require 'rubygems'
require File.dirname(__FILE__) + '/../lib/shipping_calc'
require 'yaml'
require 'rbconfig'
include ShippingCalc

# This example requires you to have a .dhl_info.yml file in your home dir to
# gather the login data. Check the Test secion in README.txt for more information.
begin
  os = Config::CONFIG["host_os"]
  if os =~ /darwin/
    auth_info = YAML.load_file("/Users/#{ENV["USER"]}/.dhl_info.yml")
    else      
    auth_info = YAML.load_file("/home/#{ENV["USER"]}/.dhl_info.yml")
  end
rescue Exception
  print "You don't have a .dhl_info.yml file in your home directory. Please
read the \"Test\" section in README.txt.\n"
  exit
end

api_user = auth_info["api_user"]
api_pwd = auth_info["api_password"]
api_key = auth_info["shipping_key"]
intl_api_key = auth_info["international_shipping_key"]
api_accnt_num = auth_info["account_num"]

opts = { 
  :api_user => api_user,
  :api_password => api_pwd,
  :shipping_key => api_key,
  :international_shipping_key => intl_api_key,
  :account_num => api_accnt_num,
  :date => Time.now,
  :service_code => 'IE',
  :shipment_code => 'O',
  :weight => 1, # weight in lbs
  :to_street => '1533-1 Ooka',
  :to_city => 'Numazu City',
  :to_zip => '410-0022',
  :to_state => 'Shizuoka',
  :to_country => 'JP',
  :international => true
}

d = DHL.new
q = d.quote(opts)
puts "Quote: " << q.to_s << "\n"


lib/shipping_calc/ups.rb
# Copyright (c) 2008 Federico Builes

# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# 'Software'), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:

# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
# IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
# CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

require 'rubygems'
require 'rexml/document'
require 'net/http'
require 'hpricot'
include REXML

module ShippingCalc
  class UPS
    attr_accessor :test

    # Map for UPS shipping code numbers to text we should display
    SERVICE_CODES = {
        '01' => 'UPS Next Day Air',
        '02' => 'UPS Second Day Air',
        '03' => 'UPS Ground',
        '07' => 'UPS Worldwide Express (1 Day, includes brokerage fees)',
        '08' => 'UPS Worldwide Expedited (2-3 Days, includes brokerage fees)',
        '11' => 'UPS Standard (Up to 7 days, brokerage fees NOT included)',
        '12' => 'UPS 3 Day Select',
        '13' => 'UPS Next Day Air Saver',
        '14' => 'UPS Next Day Air Early A.M.',
        '54' => 'UPS Worldwide Express Plus (1 Day - Early Morning, includes brokerage fees',
        '59' => 'UPS 2nd Day Air A.M.',
        '65' => 'UPS Worldwide Saver'
    }

    def quote(params)
      @xml_pieces = []
      raise ShippingCalcError.new("Invalid parameters") if params.nil?
      raise ShippingCalcError.new("Missing shipping parameters") unless params.keys.length >= 14
      rate_estimate(params)
      make_request
    end

    # Use this to see the xml request you're sending and the xml response
    # you're receiving
    def set_debug(request_path, response_path)
        @debug = true
        @debug_request_path = request_path
        @debug_response_path = response_path
    end

    private 
    def rate_estimate(params)
      piece1 = Document.new
      piece1 << XMLDecl.new("1.0' encoding='UTF-8")
      access_request = Element.new 'AccessRequest'
      access_request.attributes['xml:lang'] = 'en-US'
      access_license_number = Element.new 'AccessLicenseNumber'
      access_license_number.text = params[:shipping_key]
      access_request << access_license_number
      user_id = Element.new 'UserId'
      user_id.text = params[:api_user]
      access_request << user_id
      password = Element.new 'Password'
      password.text = params[:api_password]
      access_request << password
      piece1 << access_request
      @xml_pieces << piece1

      piece2 = Document.new
      piece2 << XMLDecl.new("1.0' encoding='UTF-8")

      rating_service_selection_request = Element.new 'RatingServiceSelectionRequest'
      rating_service_selection_request.attributes['xml:lang'] = 'en-US'
      request = Element.new 'Request'
      transaction_reference = Element.new 'TransactionReference'
      customer_context = Element.new 'CustomerContext'
      # Note: this can be whatever we want, it just helps us synchronize
      # request/response pairs.
      customer_context.text = 'Rating and Service'
      transaction_reference << customer_context
      request << transaction_reference
      request_action = Element.new 'RequestAction'
      request_action.text = 'Rate'
      request << request_action
      request_option = Element.new 'RequestOption'
      # TODO: From the docs: "If the request does not provide an option, the
      # server defaults to rate behavior.  Rate = the server rates and
      # validates the shipment. This is the default behavior if an option is
      # not provided. Shop = The server validates the shipment, and returns
      # rates for all UPS products from the ShipFrom to the ShipTo addresses
      request_option.text = 'shop'
      request << request_option
      rating_service_selection_request << request

      pickup_type = Element.new 'PickupType'
      code = Element.new 'Code'
      # Valid codes are:
      # 01 - Daily Pickup
      # 03 - Customer Counter
      # 06 - One Time Pickup
      # 07 - On Call Air
      # 11 - Suggested Retail Rates
      # 19 - Letter Center
      # 20 - Air Service Center
      code.text = params[:pickup_code]
      pickup_type << code
      rating_service_selection_request << pickup_type

      shipment = Element.new 'Shipment'
      shipper = Element.new 'Shipper'
      address = Element.new 'Address'
      city = Element.new 'City'
      city.text = params[:from_city]
      address << city
      state_province_code = Element.new 'StateProvinceCode'
      state_province_code.text = params[:from_state]
      address << state_province_code
      country_code = Element.new 'CountryCode'
      country_code.text = params[:from_country]
      address << country_code
      postal_code = Element.new 'PostalCode'
      postal_code.text = params[:from_zip]
      address << postal_code
      shipper << address
      shipment << shipper

      ship_to = Element.new 'ShipTo'
      ship_to_address = Element.new 'Address'
      ship_to_city = Element.new 'City'
      ship_to_city.text = params[:to_city]
      ship_to_address << ship_to_city
      ship_to_state_province = Element.new 'StateProvinceCode'
      ship_to_state_province.text = params[:to_state]
      ship_to_address << ship_to_state_province
      ship_to_country_code = Element.new 'CountryCode'
      ship_to_country_code.text = params[:to_country]
      ship_to_address << ship_to_country_code
      ship_to_postal_code = Element.new 'PostalCode'
      ship_to_postal_code.text = params[:to_zip]
      ship_to_address << ship_to_postal_code
      ship_to << ship_to_address
      shipment << ship_to

      params[:package_weights].each do |package_weight|
        package = Element.new 'Package'
        packaging_type = Element.new 'PackagingType'
        code = Element.new 'Code'
        # Valid Codes are:
        # 00 = Unknown
        # 01 = UPS Letter
        # 02 = Package
        # 03 = Tube
        # 04 = Pak
        # 21 = Express Box
        # 24 = 25KG Box
        # 25 = 10KG Box
        # 30 = Pallet
        # 2a = Small Express Box
        # 2b = Medium Express Box
        # 2c = Large Express Box
        code.text = params[:package_code]
        packaging_type << code
        package << packaging_type
        xml_package_weight = Element.new 'PackageWeight'
        unit_of_measurement = Element.new 'UnitOfMeasurement'
        code = Element.new 'Code'
        # Can Either be LBS or KGS
        code.text = params[:weight_uom]
        unit_of_measurement << code
        xml_package_weight << unit_of_measurement
        weight = Element.new 'Weight'
        weight.text = package_weight
        xml_package_weight << weight
        package << xml_package_weight
        shipment << package
      end

      rating_service_selection_request << shipment

      customer_classification = Element.new 'CustomerClassification'
      code = Element.new 'Code'
      if params[:pickup_code] == '01'
          code.text = '01'
      elsif ['06', '07', '19', '20'].include? params[:pickup_code]
          code.text = '03'
      elsif params[:pickup_code] == '03'
          code.text = '04'
      else
          # Undefined in the docs...
          code.text = ''
      end
      customer_classification << code
      rating_service_selection_request << customer_classification
      
      piece2 << rating_service_selection_request
      @xml_pieces << piece2
    end

    # Sends the request to the web server and returns the response.
    def make_request
      if @test
          host = 'wwwcie.ups.com'
      else
          host = 'www.ups.com'
      end

      path = "/ups.app/xml/Rate"
      server = Net::HTTP.new(host, 443)
      data = @xml_pieces.collect{|p| p.to_s}.join("\n")
      if @debug
          File.open(@debug_request_path, 'w') do |file|
              file.puts data
          end
      end
      headers = { "Content-Type" => "text/xml"}
      server.use_ssl = true
      resp = server.post(path, data, headers)
      if @debug
          File.open(@debug_response_path, 'w') do |file|
              file.puts resp.body
          end
      end
      prices = parse_response(resp.body)
    end

    # Parses the server's response.
    def parse_response(resp)
      doc = Hpricot(resp)

      find_error_and_raise(doc) if errors_exist?(doc)

      codes = []
      doc.search('//ratedshipment/service/code') do |ele|
          codes << SERVICE_CODES[ele.innerHTML]
      end
      costs = []
      doc.search('//ratedshipment/totalcharges/monetaryValue') do |ele|
          costs << ele.innerHTML
      end

      quotes = codes.zip costs
    end

    def errors_exist?(response)
      not response.search('//responsestatuscode').innerHTML.to_i == 1
    end

    def find_error_and_raise(response)
      error_code = response.search('//errorcode')
      error_description = response.search('//errordescription')
      raise ShippingCalcError.new("UPS Error #{error_code.innerHTML.to_s}:
                                  #{error_description.innerHTML.to_s}")
    end

    def date(date)
      date ||= Time.now
      if date.kind_of?(String) && date =~ /\d{4}-\d{2}-\d{2}/ # Suppose it's valid
        return date
      end

      if date.strftime("%A") == "Sunday"    
        (date + 86400).strftime("%Y-%m-%d") # DHL doesn't ship on Sundays, add 1 day.
      else
        date.strftime("%Y-%m-%d")
      end
    end

    def state(s)
      valid_state?(s) ? s : (raise ShippingCalcError.new('Invalid state for recipient'))
    end

    def valid_state?(s)
      ShippingCalc::US_STATES.include?(s)
    end

    def zip_code(code)
      if code.class != Fixnum
        raise ShippingCalcError.new('Zip Code must be a number. Perhaps you are using a string?')
      end
      code.to_s =~ /\d{5}/ ? code.to_s : (raise ShippingCalcError.new('Invalid zip code for recipient'))
    end
  end
end
